### ISSUE TITLE
Scrollview not working properly on Lenovo Yoga 920 W10

### ISSUE THREAD

**Issue by ElliotGarbus** [kivy/kivy#6646]
### Versions

* Python: 3.73
* OS: Windows10
* Kivy: 1.11.1
* Kivy installation method: pip installed in the venv

### Description
Scrollview not working properly on Lenovo Yoga 920 W10: using the touchpad to scroll down, scrolls up.  Using a mouse wheel works as expected, using touch works as expected.  The touchpad does not scroll properly.

### Code and Logs

```python
# Spinner and Scrollview tests. Scrollup does not work properly with a trackpad.

from kivy.app import App
from kivy.lang import Builder
from kivy.metrics import dp
from kivy.uix.spinner import Spinner

kv = '''
BoxLayout:
    orientation: 'vertical'
    BoxLayout:  # Line 1
        size_hint_y: 1
        BankSpinner:
            values: ['Bank: ' + str(n) for n in range(1,31)]
            text: 'Bank Select'
    ScrollView:
        size_hint_y:3
        do_scroll_x: False
        do_scroll_y: True
        scroll_type: ['bars', 'content']
        bar_width: 20
        Label:
            size_hint_y: None
            height: self.texture_size[1]
            text_size: self.width, None
            padding: 10, 10
            text: 'really some amazing text\\n' * 100    
    
    BoxLayout:
        size_hint_y: 8
'''


class BankSpinner(Spinner):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.dropdown_cls.max_height = 10 * dp(48)

    def on_touch_up(self, touch):
        print(f'touch up:{touch}, touch.device: {touch.device} {touch.profile} ')
        if 'button' in touch.profile:
            print(f'button: {touch.button}')
        return super().on_touch_up(touch)

    def on_touch_move(self, touch):
        print(f'touch move:{touch}, touch.device: {touch.device}')
        return super().on_touch_move(touch)

    def on_touch_down(self, touch):
        print(f'touch down:{touch}, touch.device: {touch.device}')
        return super().on_touch_down(touch)


class SpinnerTestApp(App):

    def build(self):
        return Builder.load_string(kv)


SpinnerTestApp().run()
```
Additionally under Inclement's direction I instrumented window_sdl2.py
in _mainloop()  scrolling on the touchpad reports mousewheelup if the scroll gesture is up or down.  In the section that starts, 'elif action.startswith('mousewheel'):'  x and y are always reported as zero when using the scrollup or scroll down gesture on the trackpad.

The attached log is the result of opening the app, scrolling down and then scrolling up with the trackpad.
[spinnertest.txt](https://github.com/kivy/kivy/files/3965576/spinnertest.txt)

**Comment by kotano** [comment]
Same problem

**Comment by tshirtman** [comment]
When you say "using the touchpad to scroll down, scrolls up" i assume you mean "putting two fingers on the touchpad, and moving them up" (i think is now the default in windows10 as in osx?) I think we are (ironicaly) still implementing the more traditional way (of before touchscreen became ubiquitous), maybe we could check the os preference and invert the direction depending on it.

**Comment by sbsusw** [comment]
@tshirtman I think @ElliotGarbus is talking about a different problem: when using two-finger scrolling, ScrollView and Spinner widgets scroll to the bottom regardless of the direction of your gesture. Changing the os preference for scroll inversion doesn't affect this behavior.

I'm seeing this problem on a Lenovo X1 Carbon (6th gen). When I put two fingers on the touchpad and move them up, the widget scrolls down towards the bottom as expected (I use "inverted" scrolling). But when I put two fingers on the touchpad and move them down, the widget jerkily scrolls quickly up and then right back down to the bottom.

When I scroll with the middle mouse button + TrackPoint (pointing nub), the scroll behavior is normal.

Windows 10 Pro (1909), Python 3.7.1, Kivy 1.11.1

**Comment by ElliotGarbus** [comment]
@sbsusw is correct.
The ScrollView will scroll up if the gesture on the touchpad is up or down.  The same behavior exists independent of the scrolling direction in settings.  On the same machine things work correctly with a mouse.  The issue is only on the touchpad.  I do not have this issue on any other W10 machines, or the mac.  Just the Lenovo Yoga 920W10.

**Comment by tshirtman** [comment]
ah, that does ring a bell, i don't know anymore if it was with you, but i remember a discussion about this, it's a bit hard to investigate if a specific hardware is needed :/.

**Comment by ElliotGarbus** [comment]
@tshirtman If I can do anything to help, Let me know.  As I recall, @inclement thought this was an issue in SDL.

**Comment by sbsusw** [comment]
I'd be very willing to help investigate, though I don't have a clear idea of where to start.

Another potentially helpful detail: weirdly enough, if I use the two-finger pinch gesture (pinch-to-zoom), I can scroll up and down perfectly by pinching inward and outward. Using the code example in the original post, it prints out "scrollup" for an inward pinch, and "scrolldown" for an outward pinch. Two-finger scroll only ever prints out "scrollup".

To get the pinch-to-zoom to work I had to disable multi-touch:
`from kivy.config import Config`
`Config.set( 'input', 'mouse', 'mouse,disable_multitouch')`

**Comment by kylona** [comment]
Has anyone found a solution to this? I'm having the same problem on a DELL Latitude. It has to do with windows gesture support on touch pad.

**Comment by ichisadashioko** [comment]
Do we or any libraries that we depended on (e.g. `SDL2`) use something like [`RegisterRawInputDevices`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-registerrawinputdevices) on Windows? I think they might do something there.

When we have a [Windows Precision Touchpad](https://docs.microsoft.com/en-us/windows-hardware/design/component-guidelines/windows-precision-touchpad-required-hid-top-level-collection), it is registered as 4 input devices.

- Configuration
- Normal mouse input
- Multi-touch input (our code or some libraries may not handle this device properly)
- Firmware update

I just `grep` `chromium` the other day and found that they did use that for the some touch/mouse input on Windows (You can zoom in and out the web page with touchpad without changing web page layout.)

**Comment by klsc-zeat** [comment]
Same problem on different Dell Precision 55x0 devices

**Comment by ichisadashioko** [comment]
I believe that some programs made with Java will not response to scrolling gestures from the track pad on Windows at all (e.g. JD-GUI) but they will response with the mouse scrolling wheel. As we don't have control over what the implementation do under the hood, it is not clear where to start to investigate this problem.

**Comment by snuq** [comment]
seems to be the same issue as #6600 , im also having trouble with this and tried to track it down, see my comment in that thread for more details... also to illustrate the issue, this is what happens when i try to scroll up (ONLY scrolling up):
![scrollviewproblem](https://user-images.githubusercontent.com/13342757/90997789-52b9d200-e577-11ea-9cb5-e1146d155389.gif)

**Comment by bajisci** [comment]
Just commenting that 2 finger scroll works here on my macbook. Kivy 1.11

**Comment by snuq** [comment]
Still having this problem with kivy 2.0.0 on windows...

**Comment by probablyfine** [comment]
Same-- just installed 2.0.0 in a fresh virtual environment, this issue seems unchanged. Using a lenovo X1 carbon

**Comment by tshirtman** [comment]
No idea how to reproduce, on my thinkpad t470 it doesn't happen on windows or linux :/ (it didn't happen with rc4 either).

**Comment by tshirtman** [comment]
So it seems we are getting a lot of scroll events without any x/y offset, and as we are doing a really simple test in
https://github.com/kivy/kivy/blob/master/kivy/core/window/_window_sdl2.pyx#L574-L578

we consider events with x=0 and y=0 to be mousewheelup, and considering we get something like

```
mousewheelup (0.0, -2.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -2.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -2.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -2.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -2.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -2.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -3.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, 0.0, None)
mousewheelup (0.0, -1.0, None)
mousewheelup (0.0, 0.0, None)
```
the result is that for any "real" mousewheel event, (corresponding to a wheel moving one step), we get a lot of "null" events, that we consider to be scroll downs (because our terminology is a bit confusing, and "mousewheelup" actually means the content goes up, and the views goes down), so while it's really easy to scroll down (there are a lot of fake events helping us this way), it's really hard to go up (these same fake events are working against us).

A trivial fix is to simply ignore these events in the window provider, this can be done in python if you want to edit a local installation.

https://github.com/kivy/kivy/blob/master/kivy/core/window/window_sdl2.py#L555
add `if not any(x, y): return`.

There might be a more proper fix.

**Comment by probablyfine** [comment]
@tshirtman many thanks for this!! It's working great for me now. For anyone looking to make this change, note that the 'any' function expects an iterable, so the line should be: `if not any([x, y]): return`

**Comment by ElliotGarbus** [comment]
@tshirtman Your proposed fix works for me on the Lenovo Yoga 910.

**Comment by tshirtman** [comment]
Indeed the line i proposed was slightly wrong, @probablyfine is correct.
The PR will do a slightly different change to avoid dispatching the event at all, but the spirit is the same, this workaround is still fine until the fix is shipped in next release.

**Comment by jballipalli** [comment]
@tshirtman Great fix...Scrolling is working perfectly on my DELL Inspiron 5567....Hope this will be fixed in next release

**Comment by Gaysha1337** [comment]
> @tshirtman many thanks for this!! It's working great for me now. For anyone looking to make this change, note that the 'any' function expects an iterable, so the line should be: `if not any([x, y]): return`

Just confirming that this work on my Asus Zenbook running Windows 10.

**Comment by tariksetia** [comment]
I am using kivy==1.11.1 and facing the same issue. 2 fingers scroll up no working properly. 
any help is appreciated.

**Comment by probablyfine** [comment]
@tariksetia I think the same fix posted above will work for that version of kivy. You just need to find [this line](https://github.com/kivy/kivy/blob/stable-1.11.1/kivy/core/window/window_sdl2.py#L554), and immediately after add this:
```
if not any( (x, y)):
    return
```
I haven't installed that version of kivy to test but that part of the file looks very similar to the most current version, so i think it will be okay.

**Comment by tshirtman** [comment]
Yes, you should be able to change it in your installation, but it's advised to upgrade to kivy 2.0 if you can.

**Comment by tariksetia** [comment]
I am using mapview, it doesn't work well with Kivy 2.0.0. So, I will have to use version 1.11.1 and looks like adding lines code suggested by @probablyfine  fix the scroll issue.

That said, Scroll issue also exist in 2.0 (at least for me)... video [here](https://drive.google.com/file/d/11y6_NDK4AP1baA41jl_DGdcuOnHRF-Im/view). you can see the stutter when scrolling up.

**Comment by tshirtman** [comment]
It might be a different issue, if the suggested change doesn't work either, then it's probably worth opening a new issue to investigate.

**Comment by probablyfine** [comment]
I'm pretty sure the fix didn't make it into 2.0.0, but it is in the current master branch.

**Comment by tshirtman** [comment]
Oh indeed, my bad, it was merged after 2.0.0 release, and will be in 2.1.0.
